### PR TITLE
UI themes for dark-base, Dracula, Nord polar night, Nord snow storm and Arc

### LINKED ISSUE #210: [color request] custom dark theme colors (customs stylesheets?)
hi. I don't like the dark theme color can you add another color for dark themes? like arc (arc-dark) or nord or another
https://www.pling.com/p/1167639/
https://github.com/horst3180/arc-theme
https://github.com/arcticicestudio/nord

### PR DESCRIPTION
**Pull request by ronilaukkarinen**
# Description

New UI themes, separate them structure under src/ui-themes which are called in src/styles.scss.

## Issues Resolved

#210

### CODE REVIEW

**Comment by ronilaukkarinen** [comment]
Just added three more UI themes:

**Nord Polar Night**
localhost_4200_ (4)

**Nord Snow Storm**
localhost_4200_ (5)

**Arc**
localhost_4200_ (6)

**Review by johannesjo — CHANGES_REQUESTED** [review]
Thank you very much! Looks pretty slick. I made some comments for some minor adjustments and questions.

One thing I recognized is that we loose all color coding for the projects. While I understand that background gradients and background colors won't work well with the themes, I tink it would be good to leave some sort of indication to indicate on which list a user is currently. 

Screenshot from 2020-11-21 17-56-48

File: package.json [hunk 1]
```diff
@@ -177,6 +177,14 @@
     "tslint": "~6.1.2",
     "typescript": "~4.0.2",
     "webdav": "^3.5.0",
-    "zone.js": "~0.11.1"
+    "zone.js": "~0.11.1",
+    "stylelint": "^13.7.2",
+    "stylelint-config-recommended": "^3.0.0",
+    "stylelint-config-standard": "^20.0.0",
+    "stylelint-config-wordpress": "^17.0.0",
```
[the remark this answers is not in the document]
Comment (ronilaukkarinen) [reply package.json:184]: Actually stylelint-config-wordpress is useless. The rest should be there, they extend the stylelintrc parameters automatically.

File: src/styles/ui-themes/dark-base/layout/_dialogs.scss [hunk 2]
```diff
@@ -0,0 +1,35 @@
+/* stylelint-disable selector-type-no-unknown, selector-max-class, selector-max-specificity, selector-max-combinators, selector-max-compound-selectors */
+
+body,
```
[the remark this answers is not in the document]
Comment (ronilaukkarinen) [reply src/styles/ui-themes/dark-base/layout/_dialogs.scss:3]: Yeah, I'm aware, I just wasn't sure which to select. Generating CSS output would need some tweaking as well, wasn't sure where SCSS to CSS tasks are located. I myself use gulp and webpack so this is setup is not yet that familiar to me. I noticed some SCSS built-in functions like `darken` didn't work, I guess some packages missing.

File: src/styles/ui-themes/dark-base/layout/_forms.scss [hunk 3]
```diff
@@ -0,0 +1,72 @@
+/* stylelint-disable selector-type-no-unknown, selector-max-class, selector-max-specificity, selector-max-combinators, selector-max-compound-selectors */
+
+body,
+body.isDarkTheme,
+body.isLightTheme,
+body.isDarkTheme.isDisableBackgroundGradient,
+body.isLightTheme.isDisableBackgroundGradient,
+body.isMac.isEnabledBackgroundGradient,
+body.isWeb.isEnabledBackgroundGradient,
+body.isEnabledBackgroundGradient,
```
[the remark this answers is not in the document]
Comment (ronilaukkarinen) [reply src/styles/ui-themes/dark-base/layout/_forms.scss:10]: > Yeah, I'm aware, I just wasn't sure which to select. Generating CSS output would need some tweaking as well, wasn't sure where SCSS to CSS tasks are located. I myself use gulp and webpack so this is setup is not yet that familiar to me. I noticed some SCSS built-in functions like darken didn't work, I guess some packages missing.

File: src/styles/ui-themes/dark-base/layout/_task.scss [hunk 4]
```diff
@@ -0,0 +1,56 @@
+/* stylelint-disable selector-type-no-unknown, selector-max-class, selector-max-specificity, selector-max-combinators, selector-max-compound-selectors */
+
+body,
+body.isDarkTheme,
+body.isLightTheme,
```
[the remark this answers is not in the document]
Comment (ronilaukkarinen) [reply src/styles/ui-themes/dark-base/layout/_task.scss:5]: > Yeah, I'm aware, I just wasn't sure which to select. Generating CSS output would need some tweaking as well, wasn't sure where SCSS to CSS tasks are located. I myself use gulp and webpack so this is setup is not yet that familiar to me. I noticed some SCSS built-in functions like darken didn't work, I guess some packages missing.

File: src/styles/ui-themes/dark-base/layout/_top-panel.scss [hunk 5]
```diff
@@ -0,0 +1,24 @@
+/* stylelint-disable selector-type-no-unknown, selector-max-class, selector-max-specificity, selector-max-combinators, selector-max-compound-selectors */
+
+body,
+body.isDarkTheme,
+body.isLightTheme,
+body.isDarkTheme.isDisableBackgroundGradient,
+body.isLightTheme.isDisableBackgroundGradient,
+body.isMac.isEnabledBackgroundGradient,
+body.isWeb.isEnabledBackgroundGradient,
+body.isEnabledBackgroundGradient,
```
[the remark this answers is not in the document]
Comment (ronilaukkarinen) [reply src/styles/ui-themes/dark-base/layout/_top-panel.scss:10]: > Yeah, I'm aware, I just wasn't sure which to select. Generating CSS output would need some tweaking as well, wasn't sure where SCSS to CSS tasks are located. I myself use gulp and webpack so this is setup is not yet that familiar to me. I noticed some SCSS built-in functions like darken didn't work, I guess some packages missing.

**Comment by ronilaukkarinen** [comment]
> One thing I recognized is that we loose all color coding for the projects. While I understand that background gradients and background colors won't work well with the themes, I tink it would be good to leave some sort of indication to indicate on which list a user is currently.

Oops, I might have overridden them by accident.

I'll go through these as soon as I have a moment 👍

**Comment by johannesjo** [comment]
There are a couple of other issues I recognized:
Screenshot from 2020-11-21 18-46-01
Screenshot from 2020-11-21 18-45-35
Screenshot from 2020-11-21 18-45-15
Screenshot from 2020-11-21 18-44-47
Screenshot from 2020-11-21 18-44-18
Screenshot from 2020-11-21 18-44-09
Screenshot from 2020-11-21 18-43-55
Screenshot from 2020-11-21 18-43-48
Screenshot from 2020-11-21 18-43-29
Screenshot from 2020-11-21 18-43-09
Screenshot from 2020-11-21 18-42-33
Screenshot from 2020-11-21 18-41-28
Screenshot from 2020-11-21 18-49-23
Screenshot from 2020-11-21 18-49-14

**Comment by johannesjo** [comment]
I also wanted to ask you about your opinion on the best approaching how to include this. Here is an incomplete list of arguments for the two main approaches that came to my mind:

Include directly as a setting inside app and bundle them with it:
* more user friendly & easier to use

As separate css only module:
* smaller size of binaries
* much easier for people to create new themes, as they don't require any knowledge of the technologies used in the main project 
* same goes for providing bugfixes to the existing themes
* users are more likely to give us some leeway for smaller styling mistakes
* possibly ess work for me :)

As you can imagine, I'm leaning more towards the second option, but maybe I am missing something? It's probably also possible to improve the user experience for the second approach, e.g. by offering to download them for the user or something like this.

**Comment by ronilaukkarinen** [comment]
Fixed the issues presented in your screenshots. I admit I didn't have enough time to test the styles, glad you went through it! There should not be any glitches left but let me know if you spot something.

> As you can imagine, I'm leaning more towards the second option, but maybe I am missing something? It's probably also possible to improve the user experience for the second approach, e.g. by offering to download them for the user or something like this.

I agree, the second options sounds better. AFAIK there is not much to miss here, at least I can't develop any bad scenarious in my mind, so go for it! I trust your approach :) Sounds good! The best choice is to have as good as possible open source experience so it's easy just to concencrate on CSS, I stand by and endorse this attitude! 👍

**Comment by johannesjo** [comment]
> The best choice is to have as good as possible open source experience so it's easy just to concencrate on CSS, I stand by and endorse this attitude!

That's great! I can create a repository if you like or you can create it. I am ok with it either way. If you create it, maybe you can add me as a collaborator so I can merge quick fixes and stuff like that myself.

**Comment by ronilaukkarinen** [comment]
@johannesjo You have the best overall understanding about super productivity and the hunch about how this would work so go ahead and do it. I'll just follow your lead then.

**Comment by johannesjo** [comment]
Alright then :) 

I created the repository and added you as a collaborator. I hope this gives you sufficient access to create anything as needed. Please let me know if there is anything missing. I'll integrate the themes later on, once this is set up. In a first step, I will simply add a section to the readme. I hope to make the whole process more user friendly later on.

**Comment by ronilaukkarinen** [comment]
Nice! I'll move the current work there as soon as I have time again :) Just to make sure: Do I just move src/styles/ui-themes there or what do you suggest? On top of that I could write a simple guidelines to README.md on how to create a new theme. This is about clarifying roles really, but can you implement that to super-productivity theme settings somehow so I don't need to worry about that or how do we do it?

**Comment by johannesjo** [comment]
> This is about clarifying roles really, but can you implement that to super-productivity theme settings somehow so I don't need to worry about that or how do we do it?

This is up to you and how you prefer to work I'd say :) 

What I had in my head was to have a repository with all the convenience stuff needed to produce new styles. So if you like to work with pre-processors and post-processors it could have a structure like this:
```
dist/some-theme.css
src/some-theme/some-theme.scss
src/some-theme/variables.scss
package.json  // <-- includes build scripts if any
README.md 
```
But I am open to anything really.

**Comment by johannesjo** [comment]
EDIT: sorry I misread your question. Yes as you write. As soon as we have some downloadable CSS files available in the theme repository, I'll add a section to the readme about all this. Making this more convenient for the user is something we can add later on I'd say.

**Comment by johannesjo** [comment]
Well that's at least the idea I have in mind. I am open to handling things differently if you prefer something else.

**Comment by ronilaukkarinen** [comment]
Ah right, this sounds good! Let me summarize this to make sure I understood everything:

I could implement my most convenient build process for these themes (SCSS->CSS) (which is via gulp tasks, stylelint etc. included in package.json) and I add tutorial on how one could install and build themes to README.md. When this is done, you later do your magic to make this work with super-productivity repo, right?

**Comment by johannesjo** [comment]
> I could implement my most convenient build process for these themes (SCSS->CSS) (which is via gulp tasks, stylelint etc. included in package.json) and I add tutorial on how one could install and build themes to README.md. When this is done, you later do your magic to make this work with super-productivity repo, right?

This would be the ideal case, yes! I haven't thought about it before but the tutorial would be really good to have, too. 

Some warning about using gulp maybe. I am not against using it and it served me very well in the past, but I get the feeling that the project is slowly dying or at least coming out of fashion. As far as I can see people are more likely to just use npm scripts to spin up their build process these days and younger developers might not be familiar with it anymore.

**Comment by ronilaukkarinen** [comment]
As being part of the gulp family in some sense, I can assure you it's alive and well. My first task runner was grunt back in 2013, grunt was a big upgrade and its development has been consistent although lately it has been slowing down a bit (last update on GitHub is from last month). Gulp 4.0 was released last year and it was a major rewrite compared to gulp 3. Gulp has many advantages in the CSS world for example live injecting style changes without full browser reload or full recompile. Gulp complements npm, they are not mutually exclusive. I haven't yet seen people going away from gulp in my field, they have just integrated gulp with other build processes alongside webpack and npm, like me.

The biggest advantage in gulp is running tasks parallel, you just need a command `gulp` that watches the files, lints and injects changes on save. I haven't found an alternative for npm scripts yet. If you do know one good alternative way, send me a GitHub link, I'm always eager to learn new methods!

I understand that the less tools the better, but in gulp task runner world especially CSS related tasks still seem to be more well thought in my opinion. You could not easily get those things done perf-wise better in npm script world. Gulp focuses more what happens inside the browser, npm focuses from tooling to building. After all, gulp is one package and consists of js so no big deal here. Here's an example of my [styles task](https://github.com/digitoimistodude/devpackages/blob/master/gulp/tasks/styles.js) if you want to take a look.

**Comment by johannesjo** [comment]
Thank you very much for the explanation! I was not aware of some of the points you make and as you say gulp is a great tool. Let's use gulp then! Also, people probably won't need to fiddle with the build process too much anyway.

**Comment by ronilaukkarinen** [comment]
@johannesjo I have now all the files in place in [super-productivity-themes](https://github.com/johannesjo/super-productivity-themes) repository. They contain a really simple gulp setup and instructions on how to contribute. Only thing that is currently a mystery to me is _how to test themes_. I have now pointed users out to clone the [styles-dev branch of my fork](https://github.com/ronilaukkarinen/super-productivity/tree/styles-dev) of super-productivity because that's the way I'm currently testing the themes myself inside [styles.scss](https://github.com/ronilaukkarinen/super-productivity/blob/styles-dev/src/styles.scss) (I use dev version daily so I can instantly fix UI issues to styles if there's any).

The only difference to my fork of super-productivity structure are
- The directory structure is _src/themename_ instead of _src/ui-themes/themename_
- I had o rename _index.scss under themes to themename.scss (for example arc.scss) so that gulp would pick up the file name and add it to dist/ with its correspondent. (in scss underscore prefixed files are "invisible" modules or dependencies and will not be picked up by task runners or scripts as where non-underscore files are considered as full styles).
- [Included dark-base as dependency](https://github.com/johannesjo/super-productivity-themes/blob/173a2eeb5da08423357de0aa4dcadb89a5ec2822/src/dracula/dracula.scss#L2) for all themes so the outputted css files contain all that is needed. So it should work when you just include the dracula.css for example inside super-productivity. Nothing else needed. Of course we have to test this in some way.

Let me know if there's something you need from me. Can't wait to have a theme option in settings in production :)

**Comment by johannesjo** [comment]
Great job @ronilaukkarinen !! I'll try to think of a good way to implement them this weekend.

**Comment by johannesjo** [comment]
I opened up a [discussion here](https://github.com/johannesjo/super-productivity/discussions/713) so that we have a better place to discuss follow up questions. This can be closed then for now, right?

**Comment by ronilaukkarinen** [comment]
Yeah, by all means. Closed.

**Pull request closed without merging**